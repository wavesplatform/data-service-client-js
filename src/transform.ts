import { Asset, IAssetJSON, BigNumber } from '@waves/data-entities';
import { TRANSACTION_TYPE } from '@waves/ts-types';
import {
  ApiTypes,
  TCandleBase,
  TPairBase,
  TApiResponse,
  TApiElement,
  TransformationResult,
  ExchangeTransaction,
  ExchangeTransactionJSON,
} from './types';
import { id } from './utils';

const transformAsset = (data: IAssetJSON | null): Asset | null =>
  data === null ? null : new Asset(data);

export const transformPair = (
  pair: TPairBase<string | number>
): TPairBase<BigNumber> => ({
  ...pair,
  firstPrice: new BigNumber(pair.firstPrice),
  lastPrice: new BigNumber(pair.lastPrice),
  low: new BigNumber(pair.low),
  high: new BigNumber(pair.high),
  volume: new BigNumber(pair.volume),
  quoteVolume: new BigNumber(pair.quoteVolume),
  volumeWaves: new BigNumber(pair.volumeWaves),
  weightedAveragePrice: new BigNumber(pair.weightedAveragePrice),
});

const toBigNumber = (v: string | number | null) =>
  v ? new BigNumber(v) : null;

export const transformCandle = (
  candle: TCandleBase<string | number>
): TCandleBase<BigNumber | null> => ({
  ...candle,
  open: toBigNumber(candle.open),
  close: toBigNumber(candle.close),
  high: toBigNumber(candle.high),
  low: toBigNumber(candle.low),
  weightedAveragePrice: toBigNumber(candle.weightedAveragePrice),
  volume: toBigNumber(candle.volume),
  quoteVolume: toBigNumber(candle.quoteVolume),
});

export const transformExchangeTransaction = (
  data: ExchangeTransactionJSON
): ExchangeTransaction => ({
  ...data,
  fee: new BigNumber(data.fee),
  price: new BigNumber(data.price),
  amount: new BigNumber(data.amount),
  buyMatcherFee: new BigNumber(data.buyMatcherFee),
  sellMatcherFee: new BigNumber(data.sellMatcherFee),
  order1: {
    ...data.order1,
    price: new BigNumber(data.order1.price),
    amount: new BigNumber(data.order1.amount),
    matcherFee: new BigNumber(data.order1.matcherFee),
  },
  order2: {
    ...data.order2,
    price: new BigNumber(data.order2.price),
    amount: new BigNumber(data.order2.amount),
    matcherFee: new BigNumber(data.order2.matcherFee),
  },
});

const transformer = (
  response: TApiResponse
): TransformationResult | TransformationResult[] => {
  switch (response.__type) {
    case ApiTypes.List:
      return response.data.map(transformSingleElement);
    default:
      return transformSingleElement(response);
  }
};

const transformSingleElement = (el: TApiElement): TransformationResult => {
  if (el === null) return null;
  else {
    switch (el.__type) {
      case ApiTypes.Asset:
        return transformAsset(el.data);
      case ApiTypes.Alias:
        return el.data;
      case ApiTypes.Pair:
        return transformPair({
          ...el.data,
          amountAsset: el.amountAsset,
          priceAsset: el.priceAsset,
        });
      case ApiTypes.Candle:
        return transformCandle(el.data);
      case ApiTypes.Transaction:
        switch (el.data.type) {
          case TRANSACTION_TYPE.EXCHANGE:
            return transformExchangeTransaction(el.data);
          default:
            return el.data;
        }
      default:
        throw 'Transformation Error';
    }
  }
};

export default transformer;
