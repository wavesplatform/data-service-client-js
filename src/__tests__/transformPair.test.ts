import { transformPair } from '../transform';
import { BigNumber } from '@waves/data-entities';

describe('Pairs transformation', () => {
  it('should transform pair with number', () => {
    const time = new Date();
    const pair = {
      firstPrice: 1.1,
      lastPrice: 2.1,
      low: 1,
      high: 3,
      volume: 100.22,
      quoteVolume: 99.2,
      volumeWaves: 10.33,
      weightedAveragePrice: 1.22,
      txsCount: 102,
      amountAsset: 'test1',
      priceAsset: 'test2',
    };
    const newPair = transformPair(pair);
    expect(newPair.firstPrice).toBeInstanceOf(BigNumber);
    expect(newPair.lastPrice).toBeInstanceOf(BigNumber);
    expect(newPair.low).toBeInstanceOf(BigNumber);
    expect(newPair.high).toBeInstanceOf(BigNumber);
    expect(newPair.volume).toBeInstanceOf(BigNumber);
    expect(newPair.quoteVolume).toBeInstanceOf(BigNumber);
    expect(newPair.volumeWaves).toBeInstanceOf(BigNumber);
    expect(newPair.weightedAveragePrice).toBeInstanceOf(BigNumber);
    expect(typeof newPair.txsCount).toEqual('number');
  });

  it('should transform pair with string', () => {
    const time = new Date();
    const pair = {
      firstPrice: '1.1',
      lastPrice: '2.1',
      low: '1',
      high: '3',
      volume: '100.22',
      quoteVolume: '99.2',
      volumeWaves: '10.33',
      weightedAveragePrice: '1.22',
      txsCount: 102,
      amountAsset: 'test1',
      priceAsset: 'test2',
    };
    const newPair = transformPair(pair);
    expect(newPair.firstPrice).toBeInstanceOf(BigNumber);
    expect(newPair.lastPrice).toBeInstanceOf(BigNumber);
    expect(newPair.low).toBeInstanceOf(BigNumber);
    expect(newPair.high).toBeInstanceOf(BigNumber);
    expect(newPair.volume).toBeInstanceOf(BigNumber);
    expect(newPair.quoteVolume).toBeInstanceOf(BigNumber);
    expect(newPair.volumeWaves).toBeInstanceOf(BigNumber);
    expect(newPair.weightedAveragePrice).toBeInstanceOf(BigNumber);
    expect(typeof newPair.txsCount).toEqual('number');
  });

  it('should transform correct', () => {
    const time = new Date();
    const pair = {
      firstPrice: '1.1',
      lastPrice: '2.1',
      low: '1',
      high: '3',
      volume: '100.22',
      quoteVolume: '99.2',
      volumeWaves: '10.33',
      weightedAveragePrice: '1.22',
      txsCount: 102,
      amountAsset: 'test1',
      priceAsset: 'test2',
    };
    const newPair = transformPair(pair);
    expect(newPair).toMatchSnapshot();
  });
});
