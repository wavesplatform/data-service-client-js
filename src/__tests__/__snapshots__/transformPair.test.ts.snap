// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pairs transformation should transform correct 1`] = `
Object {
  "amountAsset": "test1",
  "firstPrice": "1.1",
  "high": "3",
  "lastPrice": "2.1",
  "low": "1",
  "priceAsset": "test2",
  "quoteVolume": "99.2",
  "txsCount": 102,
  "volume": "100.22",
  "volumeWaves": "10.33",
  "weightedAveragePrice": "1.22",
}
`;
